README.md
# basil

Basil is a general purpose template library for Clojure and ClojureScript.

**Important:** Consider this library in _ALPHA_ until this notice is removed.
API and implementation will change in incompatible ways. You have been warned.


## Usage

On Clojars: https://clojars.org/basil

Leiningen dependency: `[basil "0.4.1"]`

Supported Clojure versions: 1.2, 1.3, 1.4, 1.5

Tested with lein-cljsbuild version: 0.2.8

This page has only elementary documentation.
[Intro page](https://github.com/kumarshantanu/basil/blob/master/doc/intro.md)
has more details.

### Basic Templates

A Basil template can be parsed from a plain string. Every Basil template is
composed of distinct, interleaved and non-nested _static-text_ and _slots_. The
_slots_ contain limited Clojure forms to express the _filter-functions_ and
_dynamic data_. Few examples:

<table>
  <tr>
    <th>Template</th>
    <th>Clojure data</th>
    <th>Result</th>
  </tr>
  <tr>
    <td>foo bar</td>
    <td>{}</td>
    <td>foo bar</td>
  </tr>
  <tr>
    <td>foo &lt;% num %&gt; bar</td>
    <td>{:num 45}</td>
    <td>foo 45 bar</td>
  </tr>
  <tr>
    <td>foo &lt;% (inc num) %&gt; bar</td>
    <td>{:inc inc :num 45}</td>
    <td>foo 46 bar</td>
  </tr>
  <tr>
    <td>foo &lt;% (inc (inc num)) %&gt; bar</td>
    <td>{:inc inc :num 45}</td>
    <td>foo 47 bar</td>
  </tr>
  <tr>
    <td><pre>foo
&lt;% (when-not (seq names) "No names!") %&gt;
bar</pre></td>
    <td>{:names []}</td>
    <td>foo<br/>
No names!<br/>
bar</td>
  </tr>
  <tr>
    <td><pre>foo
&lt;% (when (seq names) (str-br names)) %&gt;
bar</pre></td>
    <td>{:names ["Lucy" "Brian"]}</td>
    <td>foo<br/>
Lucy&lt;br/&gt;<br/>
Brian&lt;br/&gt;<br/>
bar</td>
  </tr>
  <tr>
    <td><pre>foo
&lt;% (for-each [b names] (str "name=" b)) %&gt;
bar</pre></td>
    <td>{:names ["Tom" "Jane" "Larry"]}</td>
    <td>foo<br/>
name=Tom<br/>
name=Jane<br/>
name=Larry<br/>
bar</td>
  </tr>
  <tr>
    <td><pre>foo
&lt;% (for-each [a [1 2]
              b names]
     (str ":a=" a
          ", name=" b)) %&gt;
bar</pre></td>
    <td>{:names ["Tom" "Jane" "Larry"]}</td>
    <td>foo<br/>
:a=1, name=Tom<br/>
:a=1, name=Jane<br/>
:a=1, name=Larry<br/>
:a=2, name=Tom<br/>
:a=2, name=Jane<br/>
:a=2, name=Larry<br/>
bar</td>
  </tr>
</table>

Use `basil.public/parse-compile` to parse and compile a template and use
`basil.public/render-template` to render a compiled template.


### Template Groups

Template groups can be created from directory or classpath. A slot in one
template can include an entire template dynamically using
`<% (include "foo.basil") %>`. You may also create additional context for
included template using `<% (include "foo.basil" {:a 10 :b 20} {:items xs}) %>`.

Use `basil.public/make-group-from-directory` and
`basil.public/make-group-from-classpath` to create template groups. Use
`basil.public/render-by-name` to render a template from a group.


## Getting in touch

By e-mail: kumar.shantanu(at)gmail.com

On Twitter: [@kumarshantanu](http://twitter.com/kumarshantanu)


## License

Copyright © 2012-2013 Shantanu Kumar

Distributed under the Eclipse Public License, the same as Clojure.


CHANGES.md
#Changes and TODO


## [TODO] 2012-????-?? / 0.5.0

* [TODO] Wrap fn/evaluator invocation exceptions with Basil template info
  * originating from Quiddity (due to wrong number/type of args)
  * originating from user-supplied fns/evaluators/values
* [TODO] Nested template sections
* [TODO] Template distribution mechanism
* [TODO] Investigate abstracting out string concatenation
  * Auto-select efficient JVM/CLJS specific concatenation
  * Use java.lang.StringBuilder on the JVM
  * See https://github.com/clojure/clojurescript/wiki/Differences-from-Clojure
  * Use something like this for CLJS (see 'Host Interop' section at above URL)

```clojure
  (let [sb (goog.string.StringBuffer. "hello, ")]
    (.append sb "world")
    (.toString sb))
```
* [TODO] Filter fns for reversing HTML entities


## 2013-June-24 / 0.4.1

* Fix issue where missing file did not lead to a meaningful error message


## 2012-October-16 / 0.4.0

* Common cached-group feature
* Update dependencies
  * Quiddity 0.2.0
  * clip-test 0.2.0


## 2012-August-30 / 0.3.0

* ClojureScript support (via _clip-test_)
* Use _Quiddity_ to eval slots
* Update context built-ins (replace conditionals with Quiddity built-ins)


## 2012-August-21 / 0.2.0

* Streamlined filter-fns for HTML generation, esp collections
* Library functions for 'locals scope'
* Error reporting must throw exception by default
  * implicitly uses JVM/CLJS error-handler
* New API basil.public[-cljs]


## 2012-August-08 / 0.1.0

* Template with slots
* Parse/compile and Render are separate
* Pluggable error reporting
* Model
* Filter functions
* Slot features
    * attributes   <% name %>
    * literals     <% "Number" 1 "String" "Joe Walker" "Keyword" :male %>
    * filter calls <% (second name) %>
* Template Groups and "Include"
* JVM specific API for template groups
* Collection literals - vectors, sets, maps
* Cache compiled templates (re-load only if modified)
* HTML entities (escaping HTML to avoid XSS)


project.clj
(defproject basil "0.4.1"
  :description "A general purpose template library for Clojure"
  :url "https://github.com/kumarshantanu/basil"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[quiddity "0.2.0"]]
  :profiles {:tst {:dependencies [[clip-test "0.2.0"]]}
             :jst {:source-paths ["test"]
                   ;; Enable the lein hooks for: clean, compile, test, and jar.
                   :hooks [leiningen.cljsbuild]
                   :cljsbuild {:crossovers [basil.core   basil.error  basil.group
                                            basil.lib    basil.render basil.slot
                                            basil.types  basil.util   basil.vars
                                            basil.core-test
                                            basil.group-test
                                            basil.lib-test
                                            quiddity.core quiddity.lib
                                            clip-test.internal]
                               ;; Test command for running the unit tests
                               ;;     $ lein cljsbuild test
                               :test-commands {"unit" ["phantomjs"
                                                       "run-tests.js"]}
                               :builds {:test {:source-path "test-cljs"
                                               :compiler
                                               {:output-to "target/basil-test.js"
                                                ;; :optimizations nil
                                                :pretty-print true}}}}}
             :1.2 {:dependencies [[org.clojure/clojure "1.2.1"]]}
             :1.3 {:dependencies [[org.clojure/clojure "1.3.0"]]}
             :1.4 {:dependencies [[org.clojure/clojure "1.4.0"]]}
             :1.5 {:dependencies [[org.clojure/clojure "1.5.1"]]}}
  :aliases {"all" ["with-profile" "1.2,tst:1.3,tst:1.4,tst:1.5,tst"]
            "dev" ["with-profile" "1.5,tst,jst"]}
  :warn-on-reflection true
  :plugins [[lein-cljsbuild "0.2.8"]])
